### PR TITLE
Make lektor compatible with python3

### PR DESCRIPTION
**Pull request by bameda**

### CODE REVIEW

File: lektor/admin/modules/common.py [hunk 1]
```diff
@@ -1,3 +1,7 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+from builtins import object
```
Comment (mitsuhiko) [inline lektor/admin/modules/common.py:4]: I did not try the code yet, but what is builtins?

File: lektor/admin/modules/serve.py [hunk 2]
```diff
@@ -1,7 +1,13 @@
...
+
+from __future__ import unicode_literals
+from future import standard_library
+standard_library.install_aliases()
+from past.builtins import basestring
 import os
 import mimetypes
 import posixpath
-from cStringIO import StringIO
+from io import StringIO
```
Comment (mitsuhiko) [inline lektor/admin/modules/serve.py:10]: This seems incorrect. cStringIO operates on bytes, io.StringIO operates on unicode.  The code needs to continue to operate on bytes as it would be incorrect to do otherwise.

File: lektor/admin/utils.py [hunk 3]
```diff
@@ -1,3 +1,6 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
```
Comment (mitsuhiko) [inline lektor/admin/utils.py:3]: Please do not use unicode literals.  I will not merge code that regresses on Python 2 and unicode literals is the best way to achieve that.  Please use explicit `u` prefixes on Python 3.

File: lektor/build_programs.py [hunk 4]
```diff
@@ -170,7 +175,7 @@ def build_artifact(self, artifact):
 
     def _iter_paginated_children(self):
         total = self.source.datamodel.pagination_config.count_pages(self.source)
-        for page_num in xrange(1, total + 1):
+        for page_num in range(1, total + 1):
```
Comment (mitsuhiko) [inline lektor/build_programs.py:178]: Please do not use range() on Python 2 as it has inferior performance there.

File: lektor/builder.py [hunk 5]
```diff
@@ -219,7 +223,7 @@ def write_source_info(self, info):
         con = self.connect_to_database()
         try:
             cur = con.cursor()
-            for lang, title in info.title_i18n.iteritems():
+            for lang, title in list(info.title_i18n.items()):
```
Comment (mitsuhiko) [inline lektor/builder.py:226]: Please do not use items on Python 2 as this has inferior performance.

File: lektor/builder.py [hunk 6]
```diff
@@ -469,7 +473,7 @@ def checksum(self):
                 for filename in sorted(os.listdir(self.filename)):
                     if self.env.is_uninteresting_source_name(filename):
                         continue
-                    if isinstance(filename, unicode):
+                    if isinstance(filename, str):
```
Comment (mitsuhiko) [inline lektor/builder.py:476]: This regresses on Python 2.

File: lektor/datamodel.py [hunk 7]
```diff
@@ -57,7 +63,7 @@ def count_total_items(self, record):
     def count_pages(self, record):
         """Returns the total number of pages for the children of a record."""
         total = self.count_total_items(record)
-        return int(math.ceil(total / float(self.per_page)))
+        return int(math.ceil(old_div(total, float(self.per_page))))
```
Comment (mitsuhiko) [inline lektor/datamodel.py:66]: This is unnecessary.

File: lektor/datamodel.py [hunk 8]
```diff
@@ -249,7 +255,7 @@ def __init__(self, env, id, name_i18n, label_i18n=None,
         # also includes the system fields.  This is primarily used for
         # fast internal operations but also the admin.
         self.field_map = dict((x.name, x) for x in fields)
-        for key, (ty, opts) in system_fields.iteritems():
+        for key, (ty, opts) in list(system_fields.items()):
```
Comment (mitsuhiko) [inline lektor/datamodel.py:258]: Performance regression.

File: lektor/datamodel.py [hunk 9]
```diff
@@ -356,7 +362,7 @@ def get_child_replacements(self, record):
 
     def process_raw_data(self, raw_data, pad=None):
         rv = {}
-        for field in self.field_map.itervalues():
+        for field in list(self.field_map.values()):
```
Comment (mitsuhiko) [inline lektor/datamodel.py:365]: Performance regression.

File: lektor/datamodel.py [hunk 10]
```diff
@@ -407,7 +413,7 @@ def to_json(self, pad, record=None, alt=PRIMARY_ALT):
...
```
Comment (mitsuhiko) [inline lektor/datamodel.py:416]: Performance regression.

File: lektor/datamodel.py [hunk 11]
```diff
@@ -567,7 +573,7 @@ def iter_inis(path):
                 continue
             fn = os.path.join(path, filename)
             if os.path.isfile(fn):
-                base = filename[:-4].decode('ascii', 'replace')
+                base = filename[:-4]
```
Comment (mitsuhiko) [inline lektor/datamodel.py:576]: Incorrect behaviour on Python 2.

File: lektor/datamodel.py [hunk 12]
```diff
@@ -605,7 +611,7 @@ def create_model(model_id):
         rv[model_id] = mod = datamodel_from_data(env, model_data, parent)
         return mod
 
-    for model_id in data.keys():
+    for model_id in list(data.keys()):
```
Comment (mitsuhiko) [inline lektor/datamodel.py:614]: Performance regression.

File: lektor/datamodel.py [hunk 13]
```diff
@@ -629,10 +635,10 @@ def load_flowblocks(env):
 
 
 def add_system_field(name, **opts):
-    for key, value in opts.items():
+    for key, value in list(opts.items()):
```
Comment (mitsuhiko) [inline lektor/datamodel.py:638]: Performance regression.

File: lektor/datamodel.py [hunk 14]
```diff
@@ -629,10 +635,10 @@ def load_flowblocks(env):
...
         if key.endswith('_i18n'):
             base_key = key[:-5]
-            for lang, trans in load_i18n_block(value).iteritems():
+            for lang, trans in list(load_i18n_block(value).items()):
```
Comment (mitsuhiko) [inline lektor/datamodel.py:641]: Performance regression.

File: lektor/db.py [hunk 15]
```diff
@@ -1,3 +1,11 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+from builtins import next
```
Comment (mitsuhiko) [inline lektor/db.py:4]: builtins does not exist on Python 2 and these imports seem unnecessary or at least should be on the same line.  Why are those needed?

**Comment by mitsuhiko** [comment]
Thanks so much for submitting this unfortunately this does not seem to work for Lektor.

I started commenting on this but I just realized that there is no way I can merge this.  Too many regressions to go forward.  I'm going to close this.  Feel free to reopen a new pull request if you have something that does not regress on Python 2.

**Comment by mitsuhiko** [comment]
To add more supportive feedback on this: currently it might be tricky to port Lektor because there is just not enough test support for it.  We should probably work on improving the test suite to handle all cases (particularly what the builder does) before going about porting it.  The chance of regression is just too high.

**Comment by bameda** [comment]
Ok, I understand, you are right, more tests are needed. 

Nor am I convinced the result of this PR too.
